[Is a member of][User groups] 'Add' button functionality

carma12 opened the pull request:
The 'Add' button functionality allows associating any user group's member to a given user.

To ensure that an Alert component will be displayed after a new addition, a new interface AlertObject has been created to refer to a given alert (`const alerts = useAlerts()`) passed through props to a children component.

This PR depends on this one to be merged: https://github.com/freeipa/freeipa-webui/pull/297

carma12 commented on src/components/MemberOf/MemberOfUserGroups.tsx:
```diff
@@ -12,39 +12,33 @@ import MemberOfAddModal, { AvailableItems } from "./MemberOfAddModal";
 
-function paginate<Type>(array: Type[], page: number, perPage: number): Type[] {
+function paginate(
+  array: UserGroup[],
```
I reverted the changes and added the generic function to the `utils.ts` file.

carma12 commented on src/services/rpc.ts:
```diff
@@ -1114,6 +1114,21 @@ export const api = createApi({
+    groupAddMember: build.mutation<BatchRPCResponse, [string, string[]]>({
+      query: (payload) => {
+        const toUid = payload[0];
+        const listOfMembers = payload[1];
```
I changed the name to `addToGroups` and I altered the API call to receive three parameters as it follows:

carma12 commented:
@pvoborni - The code has been corrected based on your feedback.

pvoborni reviewed (commented):
Next round of review finished.

mreynolds389 commented:
Besides @pvoborni comments, LGTM

carma12 commented on src/utils/utils.tsx:
```diff
@@ -308,3 +308,16 @@ export const parseEmptyString = (str: string) => {
 };
+
+/**
+ * Given a starting and ending index, paginate the list elements to show
```
You are right. What about changing it into `Pagination of list elements based on page and page size`?

carma12 commented:
@pvoborni - There is one comment of yours that refers to outdated code in `rpc.ts` file and for some reason, I can't add more comments in that thread. Just a heads-up/FYI: I already changed that and now the endpoints look like this:
```ts
addToGroups: build.mutation<BatchRPCResponse, [string, string, string[]]>({
      query: (payload) => {
        const toId = payload[0];
        const type = payload[1];
        const listOfMembers = payload[2];
        const membersToAdd: Command[] = [];
        listOfMembers.map((member) => {
          const payloadItem = {
            method: "group_add_member",
            params: [[member], { [type]: toId }],
          } as Command;
          membersToAdd.push(payloadItem);
        });
        console.log("membersToAdd", membersToAdd);
        return getBatchCommand(membersToAdd, API_VERSION_BACKUP);
      },
    }),
```
Basically, I added a third parameter that refers to the `type`, so it would be adapted to the entity provided. WDYT?

pvoborni commented:
> @pvoborni - There is one comment of yours that refers to outdated code in `rpc.ts` file and for some reason, I can't add more comments in that thread. Just a heads-up/FYI: I already changed that and now the endpoints look like this:
> 
> 
> Basically, I added a third parameter that refers to the `type`, so it would be adapted to the entity provided. WDYT?

The comment might have been attached to outdate code (not sure why), but it was about this implementation and still applies fully.

carma12 commented:
> > @pvoborni - There is one comment of yours that refers to outdated code in `rpc.ts` file and for some reason, I can't add more comments in that thread. Just a heads-up/FYI: I already changed that and now the endpoints look like this:
> > Basically, I added a third parameter that refers to the `type`, so it would be adapted to the entity provided. WDYT?
> 
> The comment might have been attached to outdate code (not sure why), but it was about this implementation and still applies fully.

Understood. I have already corrected the code.

pvoborni reviewed (approved):
LGTM
